### ISSUE TITLE
rclone hangs on CLOSE when transferring large FTP files

### ISSUE THREAD

**Issue by qwwqwwq** [ncw/rclone#2912]
#### What is the problem you are having with rclone?

rclone is hanging when transferring large files from FTP to google storage, file size was about 36 GB. The files are successfully transferred to google storage, but rclone never exits.

I captured this stackdump  (below) during the hang. I believe it shows the process is hanging waiting for a response after sending the remote FTP server the CLOSE command. Sending CLOSE is nice but not necessary to complete the transfer, so maybe we could add a timeout to that after which rclone gives up on hearing back a response to the CLOSE command.

```
github.com/ncw/rclone/vendor/github.com/spf13/cobra.(*Command).Execute(0x1cff800, 0xc00003c0b8, 0x0)
        /home/travis/gopath/src/github.com/ncw/rclone/vendor/github.com/spf13/cobra/command.go:800 +0x2b
github.com/ncw/rclone/cmd.Main()
        /home/travis/gopath/src/github.com/ncw/rclone/cmd/cmd.go:482 +0x46
main.main()
        /home/travis/gopath/src/github.com/ncw/rclone/rclone.go:13 +0x20

goroutine 32 [select]:
github.com/ncw/rclone/cmd.startProgress.func2(0xc000a16cc0, 0xc0004fcf60, 0x1275688)
        /home/travis/gopath/src/github.com/ncw/rclone/cmd/progress.go:49 +0x113
created by github.com/ncw/rclone/cmd.startProgress
        /home/travis/gopath/src/github.com/ncw/rclone/cmd/progress.go:41 +0xcf

goroutine 41 [IO wait, 8 minutes]:
internal/poll.runtime_pollWait(0x7f93b0e40e30, 0x72, 0xc0000ab7f8)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/runtime/netpoll.go:173 +0x66
internal/poll.(*pollDesc).wait(0xc0000fbd18, 0x72, 0xffffffffffffff00, 0x13a1ca0, 0x1c9d570)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/internal/poll/fd_poll_runtime.go:85 +0x9a
internal/poll.(*pollDesc).waitRead(0xc0000fbd18, 0xc000341000, 0x1000, 0x1000)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/internal/poll/fd_poll_runtime.go:90 +0x3d
internal/poll.(*FD).Read(0xc0000fbd00, 0xc000341000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/internal/poll/fd_unix.go:169 +0x179
net.(*netFD).Read(0xc0000fbd00, 0xc000341000, 0x1000, 0x1000, 0x1e, 0x1cfbbe0, 0x1d14d80)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/fd_unix.go:202 +0x4f
net.(*conn).Read(0xc000183900, 0xc000341000, 0x1000, 0x1000, 0x0, 0x0, 0x0)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/net.go:177 +0x68
bufio.(*Reader).fill(0xc000257920)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/bufio/bufio.go:100 +0x106
bufio.(*Reader).ReadSlice(0xc000257920, 0xb0cd0a, 0xc0002c8b60, 0xc000130a00, 0x0, 0x0, 0x18)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/bufio/bufio.go:341 +0x36
bufio.(*Reader).ReadLine(0xc000257920, 0xc00003ae00, 0xc0000aba58, 0x43473c, 0xc00033d800, 0x300000002, 0xc00033d800)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/bufio/bufio.go:370 +0x34
net/textproto.(*Reader).readLineSlice(0xc00026a900, 0x43e7bc, 0xc0005aaf28, 0xc0000aba01, 0x43e69b, 0xc0005aaf28)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/textproto/reader.go:55 +0x6f
net/textproto.(*Reader).ReadLine(0xc00026a900, 0x0, 0x0, 0xc0000abb28, 0x43e5b9)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/textproto/reader.go:36 +0x2b
net/textproto.(*Reader).readCodeLine(0xc00026a900, 0xe2, 0x0, 0x0, 0xc0000abb90, 0x50eaef, 0xc0005aaf00, 0x0)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/textproto/reader.go:177 +0x2f
net/textproto.(*Reader).ReadResponse(0xc00026a900, 0xe2, 0x0, 0xc0000abc30, 0x4057bb, 0xc00042f200, 0x1276f88)
        /home/travis/.gimme/versions/go1.11.2.linux.amd64/src/net/textproto/reader.go:255 +0x4d
github.com/ncw/rclone/vendor/github.com/jlaffaye/ftp.(*Response).Close(0xc000602600, 0xc0000abc58, 0xc0002c8b00)
        /home/travis/gopath/src/github.com/ncw/rclone/vendor/github.com/jlaffaye/ftp/ftp.go:590 +0x89
github.com/ncw/rclone/backend/ftp.(*ftpReadCloser).Close(0xc0006d5f80, 0xc0000b4977, 0x18)
        /home/travis/gopath/src/github.com/ncw/rclone/backend/ftp/ftp.go:649 +0x37
github.com/ncw/rclone/fs/asyncreader.(*AsyncReader).Close(0xc0002c8af0, 0xc0000b4977, 0x18)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/asyncreader/asyncreader.go:286 +0x62
github.com/ncw/rclone/fs/accounting.(*Account).Close(0xc000104780, 0x0, 0x0)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/accounting/accounting.go:200 +0xe8
github.com/ncw/rclone/fs/operations.Copy(0x13b6380, 0xc00035a680, 0x0, 0xc0002c8b60, 0xc0000b4977, 0x18, 0x13b6300, 0xc000602400, 0xc000a016e0, 0x4, ...)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/operations/operations.go:297 +0x1176
github.com/ncw/rclone/fs/sync.(*syncCopyMove).pairCopyOrMove(0xc000976540, 0xc0001391c0, 0x13b6380, 0xc00035a680, 0xc000976618)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:280 +0x23c
created by github.com/ncw/rclone/fs/sync.(*syncCopyMove).startTransfers
        /home/travis/gopath/src/github.com/ncw/rclone/fs/sync/sync.go:306 +0x9f

goroutine 53 [chan receive, 30 minutes]:
github.com/ncw/rclone/fs/march.(*March).Run.func2(0xc0000fb080, 0xc00002f860, 0xc000a16cf0)
        /home/travis/gopath/src/github.com/ncw/rclone/fs/march/march.go:174 +0x4e
created by github.com/ncw/rclone/fs/march.(*March).Run
        /home/travis/gopath/src/github.com/ncw/rclone/fs/march/march.go:172 +0x240
```

#### What is your rclone version (output from `rclone version`)

rclone v1.45
- os/arch: linux/amd64
- go version: go1.11.2

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Debian stretch, 64bit

####  Which cloud storage system are you using? (eg Google Drive)

FTP -> Google Storage

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

rclone copy -P --ignore-checksum xxx:xxx  gcp:test-bucket/jq-test-rclone9

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

```
$ rclone copy -vv -P --ignore-checksum xxx:xxx  gcp:test-bucket/jq-test-rclone9
2019/01/12 00:54:20 DEBUG : rclone: Version "v1.45" starting with parameters ["rclone" "copy" "-vv" "-P" "--ignore-checksum" “xxx:xxx" "gcp:test-bucket/jq-test-rclone9"]
2019/01/12 00:54:20 DEBUG : Using config file from "/home/xxx/.config/rclone/rclone.conf"
2019/01/12 00:54:20 DEBUG : ftp://xxx.com:21/xxx: Connecting to FTP server
2019-01-12 00:54:23 DEBUG : ftp://xxx.com:21/xxx: Connecting to FTP server
2019-01-12 00:54:23 INFO  : Storage bucket btest-bucket path jq-test-rclone9/: Waiting for checks to finish
2019-01-12 00:54:23 INFO  : Storage bucket test-bucket path jq-test-rclone9/: Waiting for transfers to finish
2019-01-12 00:54:25 INFO  : xxx.csv.gz.SHA2: Copied (new)
Transferred:       36.107G / 36.107 GBytes, 100%, 161.105 kBytes/s, ETA 0s
Errors:                 0
Checks:                 0 / 0, -
Transferred:            1 / 2, 50%
Elapsed time:  65h16m44.8s
Transferring:
 * xxx.csv.gz
```

As you can see I let it run over the weekend, it hangs pretty much forever with this FTP server, which does not respond to the FTP CLOSE command.

**Comment by ncw** [comment]
> rclone is hanging when transferring large files from FTP to google storage, file size was about 36 GB. The files are successfully transferred to google storage, but rclone never exits.
> 
> I captured this stackdump (below) during the hang. I believe it shows the process is hanging waiting for a response after sending the remote FTP server the CLOSE command. Sending CLOSE is nice but not necessary to complete the transfer, so maybe we could add a timeout to that after which rclone gives up on hearing back a response to the CLOSE command.

Your analysis looks good.

Does this FTP server do the same things on smaller files too?

We could put a work-around in rclone easily enough.  To put a timeout on properly would require a patch to the upstream library I think.

Do you want to work on this?

**Comment by ncw** [comment]
> I built a patched rclone binary the other day that doesn’t send the close command at all, and confirmed it fixed the issue. I assume that is too extreme a solution to be accepted.

Ah that is good confirmation.

The timeout is happening on reading the response code from the data channel.

I think we should still call close, but what we could do is give it 60s (say) to do its stuff, and if it doesn't then

Here is a potential fix which does that.  It passes the integration tests - does it fix your issue?

https://beta.rclone.org/branch/v1.45-094-g4846d939-fix-2912-ftp-hang-beta/ (uploaded in 15-30 mins)

The code change looks like this...

```diff
diff --git a/backend/ftp/ftp.go b/backend/ftp/ftp.go
index fafe6b702..fafe7bc53 100644
--- a/backend/ftp/ftp.go
+++ b/backend/ftp/ftp.go
@@ -646,7 +646,21 @@ func (f *ftpReadCloser) Read(p []byte) (n int, err error) {
 
 // Close the FTP reader and return the connection to the pool
 func (f *ftpReadCloser) Close() error {
-	err := f.rc.Close()
+	var err error
+	errchan := make(chan error, 1)
+	go func() {
+		errchan <- f.rc.Close()
+	}()
+	// Wait for Close for up to 60 seconds
+	timer := time.NewTimer(60 * time.Second)
+	select {
+	case err = <-errchan:
+		timer.Stop()
+	case <-timer.C:
+		// if timer fired assume no error but connection dead
+		fs.Errorf(f.f, "Timeout when waiting for connection Close")
+		return nil
+	}
 	// if errors while reading or closing, dump the connection
 	if err != nil || f.err != nil {
 		_ = f.c.Quit()
```

**Comment by qwwqwwq** [comment]
Confirmed that binary works!

**Comment by ncw** [comment]
Thanks for testing :-)

I merged that to master and it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in 1.46
